### PR TITLE
Allow /dev/ to exist independently of /dev/random

### PR DESCRIPTION
**Pull request by QuzarDC**
Disentangling the directory /dev/ from the (u)random handler. The core thing this makes possible is to easily add further /dev/ devices alongside random (maybe a /dev/null, but notably /dev/fb).

The core changes are:
- /dev/ can now be read as a directory via open/readdir/rewinddir
- Added a driver for /dev/null. 
- Rename the existing fs_dev to fs_random with the vfs interface names being fs_rnd rather than fs_dev.
- Add a flag for nmmgr to indicate that the vfs is intended to be a device under /dev/.
- Add a flag for nmmgr for alias names as well as a type to hold them. These should allow for a minimal footprint while implementing multiple aliases for the same handler (random/urandom/srandom/prandom or fb0/fb1 etc)
- Separate /dev/random and /dev/urandom vfs to prevent duplicated checks against the filename.
- Set stat and fstat to return the same values in for (u)random and as character devices rather than regular files.
- Add a null check to realpath returning EINVAL per ( https://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html )

I'm aware this goes exactly against the suggestion in this comment: https://github.com/KallistiOS/KallistiOS/pull/109#issuecomment-1439355776 which was to avoid having a single VFS per-file under /dev/. The issue is that otherwise, implementing other devices effectively would result in reinventing the vfs handler system. Each device will invariably have its own way to deal with the array of file ops presented in the vfs interface. I've tried to reduce the duplication by introducing the alias flag and handler types to the name manager. Gating all the VFS via KOS_INIT though would be a much better way to go about shrinking their potential footprint.

### CODE REVIEW

File: examples/dreamcast/dev/devroot/Makefile [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#
+# Basic KallistiOS skeleton / test program
+# Copyright (C)2001-2004 Megan Potter
```
Comment (andressbarajas) [inline examples/dreamcast/dev/devroot/Makefile:3]: Update these two comment lines

File: kernel/fs/fs_null.c [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    st->st_nlink = 1;
+
+    return 0;
+}
+
+/* handler interface */
+static vfs_handler_t vh = {
+    /* Name handler */
+    {
+        "/dev/null",          /* name */
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:207]: Comment spacing issue compared to comments below

File: kernel/fs/fs_null.c [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    null_seek,
+    null_tell,
+    null_total,
+    NULL,
+    NULL,               /* ioctl */
+    NULL,               /* rename/move */
+    NULL,               /* unlink */
+    NULL,
+    NULL,               /* complete */
+    null_stat,           /* stat */
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:229]:
```suggestion
    null_stat,          /* stat */
```

File: kernel/fs/fs_random.c [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    if(!rnd_verify_hnd(fd))
+        return -1;
+
+    /* The size of /dev/urandom always returns 0 */
+    return 0;
+}
+
+
+/* Delete a file */
+static int rnd_unlink(vfs_handler_t * vfs, const char *path) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:202]:
```suggestion
static int rnd_unlink(vfs_handler_t *vfs, const char *path) {
```

File: kernel/fs/fs_random.c [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+static off_t rnd_tell(void * hnd) {
+    /* Check the handle */
+    if(!rnd_verify_hnd(hnd))
+        return -1;
+
+    return 0;
+}
+
+/* return the filesize */
+static size_t rnd_total(void * fd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:191]:
```suggestion
static size_t rnd_total(void *fd) {
```

File: kernel/fs/fs_random.c [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+    /* Check the handle */
+    if(!rnd_verify_hnd(hnd))
+        return -1;
+
+    return 0;
+}
+
+/* tell the current position in the file */
+static off_t rnd_tell(void * hnd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:182]:
```suggestion
static off_t rnd_tell(void *hnd) {
```

File: kernel/fs/fs_random.c [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+    if((fh->mode & O_MODE_MASK) != O_WRONLY && (fh->mode & O_MODE_MASK) != O_RDWR)
+        return -1;
+
+    dbglog(DBG_ERROR, "fs_random: writing entropy is not supported\n");
+    return -1;
+}
+
+
+/* Seek elsewhere in a file */
+static off_t rnd_seek(void * hnd, off_t offset, int whence) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:170]:
```suggestion
static off_t rnd_seek(void *hnd, off_t offset, int whence) {
```

File: kernel/fs/fs_random.c [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+    if((fh->mode & O_MODE_MASK) != O_RDONLY && (fh->mode & O_MODE_MASK) != O_RDWR)
+        return 0;
+
+    arc4random_buf(buf, cnt);
+
+    return cnt;
+}
+
+/* write function */
+static ssize_t rnd_write(void * hnd, const void *buffer, size_t cnt) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:148]:
```suggestion
static ssize_t rnd_write(void *hnd, const void *buffer, size_t cnt) {
```

File: kernel/fs/fs_random.c [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+    mutex_lock(&fh_mutex);
+    TAILQ_REMOVE(&rnd_fh, fh, listent);
+    mutex_unlock(&fh_mutex);
+
+    free(fh);
+    return retval;
+}
+
+/* read function */
+static ssize_t rnd_read(void * hnd, void *buffer, size_t cnt) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:128]:
```suggestion
static ssize_t rnd_read(void *hnd, void *buffer, size_t cnt) {
```

File: kernel/fs/fs_random.c [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+            break;
+        }
+    }
+    mutex_unlock(&fh_mutex);
+
+    return rv;
+}
+
+/* close a file */
+static int rnd_close(void * hnd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:106]:
```suggestion
static int rnd_close(void *hnd) {
```

File: kernel/fs/fs_random.c [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+    /* link the fh onto the top of the list */
+    mutex_lock(&fh_mutex);
+    TAILQ_INSERT_TAIL(&rnd_fh, fh, listent);
+    mutex_unlock(&fh_mutex);
+
+    return (void *)fh;
+}
+
+/* Verify that a given hnd is actually in the list */
+static int rnd_verify_hnd(void * hnd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:89]:
```suggestion
static int rnd_verify_hnd(void *hnd) {
```

File: kernel/fs/fs_random.c [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+    if(realmode != O_RDONLY) {
+        free(fd);
+        return NULL;
+    }
+
+    return fd;
+}
+
+/* open function */
+static void * rnd_open(vfs_handler_t * vfs, const char *path, int mode) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:74]:
```suggestion
static void * rnd_open(vfs_handler_t *vfs, const char *path, int mode) {
```

File: kernel/fs/fs_random.c [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+} rnd_fh_t;
+
+/* Linked list of open files (controlled by "mutex") */
+TAILQ_HEAD(rnd_fh_list, rnd_fh_str) rnd_fh;
+
+/* Thread mutex for rnd_fh access */
+static mutex_t fh_mutex;
+
+/* openfile function */
+static rnd_fh_t *rnd_open_file(vfs_handler_t * vfs, const char *fn, int mode) {
```
Comment (andressbarajas) [inline kernel/fs/fs_random.c:49]:
```suggestion
static rnd_fh_t *rnd_open_file(vfs_handler_t *vfs, const char *fn, int mode) {
```

File: kernel/fs/fs_null.c [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+static off_t null_tell(void * hnd) {
+    /* Check the handle */
+    if(!null_verify_hnd(hnd))
+        return -1;
+
+    return 0;
+}
+
+/* return the filesize */
+static size_t null_total(void * fd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:166]:
```suggestion
static size_t null_total(void *fd) {
```

File: kernel/fs/fs_null.c [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+    /* Check the handle */
+    if(!null_verify_hnd(hnd))
+        return -1;
+
+    return 0;
+}
+
+/* tell the current position in the file */
+static off_t null_tell(void * hnd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:157]:
```suggestion
static off_t null_tell(void *hnd) {
```

File: kernel/fs/fs_null.c [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+    /* Make sure we're opened for writing */
+    if((fh->mode & O_MODE_MASK) != O_WRONLY && (fh->mode & O_MODE_MASK) != O_RDWR)
+        return -1;
+
+    return cnt;
+}
+
+/* Seek elsewhere in a file */
+static off_t null_seek(void * hnd, off_t offset, int whence) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:145]:
```suggestion
static off_t null_seek(void *hnd, off_t offset, int whence) {
```

File: kernel/fs/fs_null.c [hunk 17]
```diff
@@ -0,0 +1,272 @@
...
+
+    /* make sure we're opened for reading */
+    if((fh->mode & O_MODE_MASK) != O_RDONLY && (fh->mode & O_MODE_MASK) != O_RDWR)
+        return 0;
+
+    return 0;
+}
+
+/* write function */
+static ssize_t null_write(void * hnd, const void *buffer, size_t cnt) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:124]:
```suggestion
static ssize_t null_write(void *hnd, const void *buffer, size_t cnt) {
```

File: kernel/fs/fs_null.c [hunk 18]
```diff
@@ -0,0 +1,272 @@
...
+    mutex_lock(&fh_mutex);
+    TAILQ_REMOVE(&null_fh, fh, listent);
+    mutex_unlock(&fh_mutex);
+
+    free(fh);
+    return 0;
+}
+
+/* read function */
+static ssize_t null_read(void * hnd, void *buffer, size_t cnt) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:102]:
```suggestion
static ssize_t null_read(void *hnd, void *buffer, size_t cnt) {
```

File: kernel/fs/fs_null.c [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+            break;
+        }
+    }
+    mutex_unlock(&fh_mutex);
+
+    return rv;
+}
+
+/* close a file */
+static int null_close(void * hnd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:81]:
```suggestion
static int null_close(void *hnd) {
```

File: kernel/fs/fs_null.c [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+    /* link the fh onto the top of the list */
+    mutex_lock(&fh_mutex);
+    TAILQ_INSERT_TAIL(&null_fh, fh, listent);
+    mutex_unlock(&fh_mutex);
+
+    return (void *)fh;
+}
+
+/* Verify that a given hnd is actually in the list */
+static int null_verify_hnd(void * hnd) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:64]:
```suggestion
static int null_verify_hnd(void *hnd) {
```

File: kernel/fs/fs_null.c [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+    fd = malloc(sizeof(null_fh_t));
+
+    /* Fill in the filehandle struct */
+    fd->mode = mode;
+
+    return fd;
+}
+
+/* open function */
+static void * null_open(vfs_handler_t * vfs, const char *path, int mode) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:49]:
```suggestion
static void * null_open(vfs_handler_t *vfs, const char *path, int mode) {
```

File: kernel/fs/fs_null.c [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+} null_fh_t;
+
+/* Linked list of open files (controlled by "mutex") */
+TAILQ_HEAD(null_fh_list, null_fh_str) null_fh;
+
+/* Thread mutex for null_fh access */
+static mutex_t fh_mutex;
+
+/* openfile function */
+static null_fh_t *null_open_file(vfs_handler_t * vfs, const char *fn, int mode) {
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:33]:
```suggestion
static null_fh_t *null_open_file(vfs_handler_t *vfs, const char *fn, int mode) {
```

File: kernel/fs/fs_null.c [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+#include <errno.h>
+#include <arch/types.h>
+#include <kos/mutex.h>
+#include <kos/fs_null.h>
+#include <sys/queue.h>
+#include <errno.h>
+
+/* File handles */
+typedef struct null_fh_str {
+    int mode;                           /* mode the file was opened with */
```
Comment (andressbarajas) [inline kernel/fs/fs_null.c:21]:
```suggestion
    int mode;                            /* mode the file was opened with */
```

**Review by ljsebald — APPROVED** [review]

**Pull request merged**